docs/architecture/adr/0022-server-authorization.md

```diff
@@ -0,0 +1,244 @@
+
+#### Advantages
+
+- strong conceptual foundation - if we go through the exercise of defining our authorization logic
+  in the OpenFGA model, we will have a single source of truth and hopefully draw out any issues or
+  inconsistencies in our current structure
+- designed for complex permission structures, e.g. supports cascading permissions (user has a view
+  relationship with a cipher if they have a Can Edit relationship with the collection the cipher is
+  in)
+- focuses on defining the model upfront, then the checks are extremely simple (because it's just
+  checking the existence of relationships). This is expected to result in minimal duplication of
+  logic given that you define a single consistent model, rather than defining many authorization
+  handlers (for example)
+- queries are more flexible and performant and are expected to cover all our use cases:
+  - no need to fetch the resource from the database in advance, just pass ids
+  - it could totally replace authorization logic in the database; instead of performing table joins
+    to determine what ciphers a user has access to, the server would query OpenFGA with a "list
+    objects request", then fetch the ciphers from the database by their ids only
+- a solution for both client and server. Clients could also query the OpenFGA store to determine
+  what UI flows are available to the user, without having to duplicate logic
+- could enable more powerful reporting/dashboard features and logging/audit trails in our products
+  because we can easily determine at any time who has access to what. This is a pain point today as
+  new reports can require complex database queries that necessitate a detailed understanding of how
+  user permissions are represented in our relational database
+- immutable, versioned authorization models allow for graceful changes to and between authorization
+  structures
+
+#### Disadvantages
+
+- additional external dependency that is not easily replaced once installed. However, it is
+  supported by the Cloud Native Computing Foundation (itself part of the Linux Foundation), see also
+  its [list of adopters](https://github.com/openfga/community/blob/main/ADOPTERS.md)
+- new concepts and a DSL unfamiliar to most developers - would probably require a single team to
+  become experts in this and maintain it for all teams
+- heavy implementation impact on engineering resources:
+  - requires remodelling of our authorization logic around OpenFGA concepts
+  - BRE/SRE resources to support the OpenFGA deployment and separate database
+  - rewrite not just authorization checks, but any write operations (commands) to ensure that the
+    OpenFGA store is always kept in step with the Bitwarden database
+- probably not compatible with Bitwarden Unified due to resource constraints (assumption only) -
+  unclear what would replace it in this case (more research required if we are considering this
+  option)
+
+### Casbin
+
+// TODO
```

Comment:
💭 on their flexible storage model being easier and therefore an advantage for self-host.